### PR TITLE
Authenticate

### PR DESCRIPTION
**Pull request by gstone271**
* Add login and heartbeat APIs
* Change DB type on SessionKeys
* Merge Elijah's in-progress db connection code
* Add configuration system for db creds, other settings
* Improved error handling
* Add basic API testing

Authentication APIs not done: logout, change password, reset password, create user

### CODE REVIEW

File: api/handler.ts [hunk 1]
```diff
@@ -1,40 +1,23 @@
...
-          "data": data
-        })
-      ).catch(ex => handleErrors(request, response, ex));
-    } catch (ex) {
-      //currently this doesn't handle errors differently for authenticated users than unauthenticated ones; we may want to ultimately hide some information from unauthenticated users
-      handleErrors(request, response, ex);
-    }
-    return undefined;
-  }
+export default function handler(perform, authenticate = auth.alwaysPermitted, handleErrors = errors.defaultErrorHandler) {
```
Comment (ElijahRW) [inline api/handler.ts:8]: The base handler is set to alwaysPermitted by default. I think our default should be to require authentication and manually expose the open API's, such as 'login'.

File: api/tests/login.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+curl localhost:3333/heartbeat -H "X-SESSION-ID: $key"
+:
+curl localhost:3333/heartbeat -H "X-USER-ID: 111"
+:
+curl localhost:3333/heartbeat -H "X-USER-ID: 54321" -H "X-SESSION-ID: 4321"
+:
+curl localhost:3333/heartbeat -H "X-USER-ID: 222" -H "X-SESSION-ID: $key"
+:
+curl localhost:3333/heartbeat -H "X-USER-ID: 111" -H "X-SESSION-ID: $key"
+:
```
Comment (ElijahRW) [inline api/tests/login.sh:25]: Clarification: Should we assume this login test passes if the script runs to completion?

**Review by ElijahRW — APPROVED** [review]
Looks Great, excellent work. I think it would be useful if you could add some brief documentation for the correct process of the Application _(IE: to grab a key: post to `URL/login`.)_  However, a single read through of the code is MORE than self explanatory.

File: api/handler.ts (hunk 1 above)
Comment (gstone271) [reply api/handler.ts:8]: You are correct, I have it set to alwaysPermitted only until our client functions know how to properly authenticate.

File: api/tests/login.sh (hunk 2 above)
Comment (gstone271) [reply api/tests/login.sh:25]: No, currently you must manually inspect the output to ensure it is as expected.

**Pull request merged**